crates/context/src/handlers/join_context.rs

```diff
@@ -0,0 +1,410 @@
+                Hash::default(),
+            ))
+        },
+        |meta| {
+            let context = Context::new(
+                context_id,
+                application_id.unwrap_or_else(|| meta.application.application_id()),
+                meta.root_hash.into(),
+            );
+
+            save_context(datastore, &context)?;
+
+            Ok(context)
+        },
+    )
+}
+
+pub async fn install_application_from_url(
+    datastore: &Store,
+    node_client: &NodeClient,
+    url: Url,
+    metadata: Vec<u8>,
+    expected_hash: Option<Hash>,
+) -> eyre::Result<ApplicationId> {
+    let uri = url.as_str().parse()?;
+
+    let response = Client::new().get(url).send().await?;
+
+    let expected_size = response.content_length();
+
+    let (blob_id, size) = add_blob(
+        node_client,
+        response
+            .bytes_stream()
+            .map_err(Error::other)
+            .into_async_read(),
+        expected_size,
+        expected_hash,
+    )
+    .await?;
+
+    install_application(datastore, blob_id, size, &uri, metadata)
+}
+
+pub async fn add_blob<S: AsyncRead>(
+    node_client: &NodeClient,
+    stream: S,
+    expected_size: Option<u64>,
+    expected_hash: Option<Hash>,
+) -> eyre::Result<(BlobId, u64)> {
+    let (blob_id, size) = node_client
+        .add_blob(stream, expected_size, expected_hash.as_ref())
+        .await?;
+
+    if matches!(expected_size, Some(expected_size) if size != expected_size) {
+        eyre::bail!("fatal: blob size mismatch");
+    }
+
+    Ok((blob_id, size))
+}
+
+fn install_application(
+    datastore: &Store,
+    blob_id: BlobId,
+    size: u64,
+    source: &ApplicationSource,
+    metadata: Vec<u8>,
+) -> eyre::Result<ApplicationId> {
+    let application = types::ApplicationMeta::new(
+        key::BlobMeta::new(blob_id),
+        size,
+        source.to_string().into_boxed_str(),
+        metadata.into_boxed_slice(),
+    );
+
+    let application_id = ApplicationId::from(*Hash::hash_borsh(&application)?);
+
+    let mut handle = datastore.handle();
+
+    handle.put(&key::ApplicationMeta::new(application_id), &application)?;
+
+    Ok(application_id)
+}
+
+pub fn is_application_installed(
+    datastore: &Store,
+    node_client: &NodeClient,
+    application_id: &ApplicationId,
+) -> eyre::Result<bool> {
+    let handle = datastore.handle();
+
+    if let Some(application) = handle.get(&key::ApplicationMeta::new(*application_id))? {
+        if has_blob_available(node_client, &application.blob.blob_id())? {
+            return Ok(true);
+        }
+    }
+
+    Ok(false)
+}
+
+pub fn has_blob_available(node_client: &NodeClient, blob_id: &BlobId) -> eyre::Result<bool> {
+    node_client.has_blob(blob_id)
+}
+
+async fn subscribe(network_client: &NodeClient, context_id: &ContextId) -> eyre::Result<()> {
+    network_client.subscribe(context_id).await?;
+
+    info!(%context_id, "Subscribed to context");
+
+    Ok(())
+}
```

Comment:
We shouldn't need the method here, it's already accounted for

https://github.com/calimero-network/core/blob/fd6c5d2ae98659eedb40e342514aa889c62f26f3/crates/node/primitives/src/client.rs#L24-42